---
title: "mIVD_scRNAseq"
author: "Jian"
date: '2024-06-15'
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=T, warning=T, cache = FALSE, cache.lazy = FALSE, dev = "png",dpi = 300)
Sys.setenv(LANGUAGE = "en") 
options(stringsAsFactors = F, future.fork.enable = TRUE, future.globals.maxSize = 2000 * 1024^1024, 
        repos = c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"),
        BioC_mirror="https://mirrors.tuna.tsinghua.edu.cn/bioconductor/")
```

# 1. Preparations
## 1.1 prepare the envs and path
```{r}
rm(list = ls())
gc()
knitr::opts_knit$set(root.dir = "~/project/mouse_IVD/")
```

## 1.2 load packages
```{r echo=TRUE, message=FALSE, warning=FALSE}
library(Seurat)
library(ggplot2)
library(patchwork)
library(pheatmap)
library(cowplot)
library(dplyr)
library(reshape2)
library(ggpubr)
library(stringr)
library(scCustomize)
```

## 1.3 load data
```{r}
file_names <- c("IVDH","IVDF")

seu_list <- lapply(file_names, function(x) {
  count <- Read10X(paste0("./counts/",x,"/filter_matrix/"), gene.column = 1)
  colnames(count) <- paste0(x, "_", colnames(count))
  seu_obj <- CreateSeuratObject(counts = count, project = x, min.features = 10)
  return(seu_obj)
})

seu <- Merge_Seurat_List(seu_list)

seu <- Add_Mito_Ribo_Seurat(seu, species = "mouse")
seu <- Add_Cell_Complexity_Seurat(seurat_object = seu)
seu$orig.ident <- factor(seu$orig.ident, levels = c("IVDH","IVDF"))
saveRDS(seu, file = "./rds/seu_raw.Rds")
rm(seu_list); gc()
```


## 1.4 QC 
```{r, fig.width=10, fig.height=3.5}
p1 <- QC_Plots_Genes(seu, low_cutoff = 1000, high_cutoff = 8000, pt.size = 0, group.by = "orig.ident",
                     colors_use = c('#a5dff9','#FFBC42'), raster = T, plot_median = TRUE)
p2 <- QC_Plots_UMIs(seu, low_cutoff = 1000, high_cutoff = 50000, pt.size = 0, group.by = "orig.ident",
                    colors_use = c('#a5dff9','#FFBC42'), raster = T, plot_median = TRUE)
p3 <- QC_Plots_Mito(seu, high_cutoff = 10, pt.size = 0, group.by = "orig.ident",
                    colors_use = c('#a5dff9','#FFBC42'), raster = T, plot_median = TRUE)
p4 <- QC_Plots_Complexity(seu, high_cutoff = 0.8, pt.size = 0, group.by = "orig.ident",
                          colors_use = c('#a5dff9','#FFBC42'), raster = T, plot_median = TRUE)

p_sample <- wrap_plots(p1, p2, p3, p4, ncol = 4)
p_sample

ggsave(p_sample, filename = "./pdf/sample_qc.pdf", width = 15, height = 4.5)
```

## 1.5 filter cells
```{r}
seu <- readRDS(file = "./rds/seu_raw.Rds")

sample_names <- c("IVDH","IVDF")
library(SCP)
seu_list <- lapply(sample_names, function(x) {
  seu_obj <- subset(seu, orig.ident == x)
  seu_obj <- RunCellQC(seu_obj, return_filtered = F, db_method = "scDblFinder", 
                       UMI_threshold = 1000, mito_threshold = 10,  
                       gene_threshold = 1000)
  return(seu_obj)
})

seu <- Merge_Seurat_List(seu_list)
table(seu$CellQC, seu$orig.ident)

seu <- subset(seu, CellQC == "Pass")
rm(seu_list); gc()
```


# 2. preprocessing
## 2.1 batch effect removal
```{r}
library(bbknnR)
seu <- NormalizeData(seu)
seu <- FindVariableFeatures(seu)
seu <- ScaleData(seu)
seu <- RunPCA(seu, npcs = 50)
seu_bb <- RunBBKNN(seu, batch_key = "orig.ident", run_UMAP = T, run_TSNE = T)
DimPlot_scCustom(seu_bb, group.by = c("orig.ident"), reduction = "umap", pt.size = 0.5,
                       shuffle = T, raster = T, colors_use = c('#a5dff9','#FFBC42'))

ggsave(filename = "./pdf/umap_sample.pdf",plot = last_plot(), width = 6, height = 4.5)
```

## 2.2 clustering
```{r, fig.width=12, fig.height=8}
seu_bb <- FindNeighbors(seu_bb, reduction = "pca", k.param = 10)
seu_bb <- FindClusters(seu_bb, resolution = 0.3, graph.name = "bbknn")

seu_c4 <- subset(seu_bb, seurat_clusters == 4)
seu_c4 <- FindNeighbors(seu_c4, dims = 1:20, k.param = 10, reduction = "pca")
seu_c4 <- FindClusters(seu_c4, resolution = 0.1, graph.name = "bbknn")
seu_bb$sub_cluster <- as.character(seu_bb$seurat_clusters)
seu_bb$sub_cluster[Cells(seu_c4)] <- paste("c4",Idents(seu_c4), sep = "_")

seu_c6 <- subset(seu_bb, seurat_clusters == 6)
seu_c6 <- FindNeighbors(seu_c6, dims = 1:20, k.param = 10, reduction = "pca")
seu_c6 <- FindClusters(seu_c6, resolution = 0.1, graph.name = "bbknn")
seu_bb$sub_cluster[Cells(seu_c6)] <- paste("c6",Idents(seu_c6), sep = "_")

DimPlot(seu_bb, group.by = c("sub_cluster"), reduction = "umap", 
                       shuffle = T, raster = T)


p1 <- Plot_Density_Joint_Only(seu_bb, features = c("T", "Cd24a", "Krt8", "Acan"),
                        reduction = "umap", pt.size = 0.1) + 
  scale_color_viridis_c(option = "inferno", direction = -1)

p2 <- Plot_Density_Joint_Only(seu_bb, features = c("Col2a1", "Sox9", "Comp","Bgn"), 
                        reduction = "umap", pt.size = 0.1)+
  scale_color_viridis_c(option = "inferno", direction = -1)

p1 /p2

# markers_AF <- FindMarkers(object = seu_bb, ident.1 = c('RegC',  'HyperC', "FibroC"), ident.2 =  'NC', 
#                           logfc.threshold = 1, min.pct = 0.25, only.pos = T)
# 
ggsave(filename = "./pdf/umap_NP_AF_sig.pdf",plot = last_plot(), width = 4.5, height = 7)
```

## 2.4 annotation
```{r}
Idents(seu_bb) <- "sub_cluster"
seu_bb <- RenameIdents(object = seu_bb, '0' = 'RegC', '1' = 'HyperC', '2' = "NC", '3' = "FibroC", 
                       'c4_0' = "CycC", 'c4_1' = "Blood", 'c4_2' = "Blood",'c4_3' = "Blood",
                       '5' = "OP",'c6_0' = "PC", "c6_1" = "Endothelium")
seu_bb$cluster <- Idents(seu_bb)

seu_bb$cluster <- factor(seu_bb$cluster, levels = c("NC",'RegC','HyperC',"FibroC", "CycC", "OP","PC",
                                                    "Endothelium","Blood"))

cluster_cols <- DiscretePalette_scCustomize(num_colors = 30, palette = "varibow")


ph_colors <- list(orig.ident = setNames(c('#a5dff9','#FFBC42'), c("IVDH","IVDF")),
                  cluster = setNames(cluster_cols[c(1,3,5,9,11,19,26,22,18)], levels(seu_bb$cluster)))
DimPlot_scCustom(seu_bb, group.by = c("cluster"), reduction = "umap", pt.size = 0.75,
                       shuffle = T, raster = T, colors_use = ph_colors$cluster, label = T)

```


# 3. cross-species correlation
## 3.1 load human dataset
```{r}
## 2022 iScience young vs adult IVD
Y1 <- Read10X("~/project/scRNA_data_2023_iScience/C01/", )
Y2 <- Read10X("~/project/scRNA_data_2023_iScience/C02/")
colnames(Y1) <- paste("Y1", colnames(Y1), sep = "_")
colnames(Y2) <- paste("Y2", colnames(Y2), sep = "_")

A1 <- Read10X("~/project/scRNA_data_2023_iScience/R01/")
A3 <- Read10X("~/project/scRNA_data_2023_iScience/R03/")
colnames(A1) <- paste("A1", colnames(A1), sep = "_")
colnames(A3) <- paste("A3", colnames(A3), sep = "_")

counts <- cbind(Y1, Y2, A1, A3)
seu_hum <- CreateSeuratObject(counts = counts, project = "human_iScience", min.cells = 3)
table(seu_hum$orig.ident)
seu_hum$group <- seu_hum$orig.ident
seu_hum$group <- gsub("A1|A3", "Adole", seu_hum$group)
seu_hum$group <- gsub("Y1|Y2", "Young", seu_hum$group)
seu_hum$age <- seu_hum$orig.ident
seu_hum$age <- gsub("Y1", "7", seu_hum$age)
seu_hum$age <- gsub("Y2", "6", seu_hum$age)
seu_hum$age <- gsub("A1", "19", seu_hum$age)
seu_hum$age <- gsub("A3", "15", seu_hum$age)
table(seu_hum$group, seu_hum$age)
saveRDS(seu_hum, file = "~/project/scRNA_data_2022_iScience/seu_orig.Rds")
```

## 3.2 re-analysis
```{r, fig.width=12, fig.height=4.5}
sample_names <- unique(seu_hum$orig.ident)
library(SCP)
seu_list <- lapply(sample_names, function(x) {
  seu_obj <- subset(seu_hum, orig.ident == x)
  seu_obj <- RunCellQC(seu_obj, return_filtered = F, db_method = "scDblFinder", 
                       UMI_threshold = 1000, mito_threshold = 10,  
                       gene_threshold = 1000)
  return(seu_obj)
})

seu_hum <- Merge_Seurat_List(seu_list)
table(seu_hum$CellQC, seu_hum$orig.ident)

seu_hum <- subset(seu_hum, CellQC == "Pass")

## batch effect removal
library(bbknnR)
seu_hum <- NormalizeData(seu_hum)
seu_hum <- FindVariableFeatures(seu_hum)
seu_hum <- ScaleData(seu_hum)
seu_hum <- RunPCA(seu_hum, npcs = 50)
seu_hum <- RunBBKNN(seu_hum, batch_key = "orig.ident", run_UMAP = T, run_TSNE = F)
DimPlot_scCustom(seu_hum, group.by = c("orig.ident"), reduction = "umap", pt.size = 0.5,
                       shuffle = T, raster = T)

seu_hum <- FindNeighbors(seu_hum, reduction = "pca", k.param = 10, dims = 1:30)
seu_hum <- FindClusters(seu_hum, resolution = 0.6, graph.name = "bbknn", algorithm = 2)


DimPlot(seu_hum, group.by = c("seurat_clusters", "group"), ncol = 2, reduction = "umap",
        cols = hej_colors, label = T)


seu_hum <- RenameIdents(object = seu_hum, '0' = 'HomC', '1' = 'HyperC', '2' = "HomC", 
                         '3' = "EffectC",'4' = "EffectC", '5' = "HyperC", "6" = "FibroC",
                         "7" = "MCAM_Prog", "8"= "Blood", "9" = "CD24_Prog", "10" = "T_cell", 
                        "11" = "Endothelium", "12" = "Mki67_Prog", "13" = "Erythrocyte")
seu_hum$cluster <- Idents(seu_hum)
seu_hum$cluster <- factor(seu_hum$cluster, levels = c('HomC', 'HyperC', "EffectC",
                                                      "FibroC","CD24_Prog","MCAM_Prog",  
                                                      "Mki67_Prog","Endothelium", 
                                                       "Blood", "T_cell", "Erythrocyte"))
saveRDS(seu_hum, file = "~/project/scRNA_data_2022_iScience/seu_bbknn.Rds")
```

## 3.3 singleR prediction
```{r, fig.height=7, fig.width=8}
#### singleR
library(SingleR)
library(BiocParallel)
library(future)
require(ggplotify)

seu_bb$cluster2 <- paste0(seu_bb$orig.ident, "_", seu_bb$cluster)

## orthologs 
library(gprofiler2)
genes.map <- gorth(rownames(seu_hum), source_organism="hsapiens", target_organism="mmusculus",
                   numeric_ns="ENTREZGENE_ACC")
counts_hum <- GetAssay(seu_hum, assay = "RNA")
counts_hum <- counts_hum@counts
gene_info <- data.frame(row.names = rownames(counts_hum),
                        input = rownames(counts_hum)) 
gene_info$orth <- plyr::mapvalues(rownames(gene_info), 
                                  genes.map$input, genes.map$ortholog_name)
gene_info <- gene_info[!duplicated(gene_info$orth), ]
counts_hum <- counts_hum[gene_info$input, ]
rownames(counts_hum) <- gene_info$orth

seu_hum2 <- CreateSeuratObject(counts = counts_hum, meta.data = seu_hum@meta.data)
seu_hum2 <- NormalizeData(seu_hum2)

gene_int <- intersect(rownames(seu_hum2), rownames(seu_bb))

scRNA_pred <- SingleR(test = seu_hum2@assays$RNA@data,
                      ref = seu_bb@assays$RNA@data,
                      labels = seu_bb$cluster,
                      de.method="wilcox",
                      BPPARAM=MulticoreParam(20))


seu_hum2 <- AddMetaData(seu_hum2, 
                        metadata = scRNA_pred[["pruned.labels"]], 
                        col.name = "singleR_predicted")


ph_colors_sr <- list(cluster = setNames(louvain_colors, unique(seu_hum2$cluster)),
                     group = setNames(c("blue","orange"), c("Young", "Adole")),
                     singleR_predicted = ph_colors$cluster)

seu_hum2$singleR_predicted <- factor(seu_hum2$singleR_predicted, 
                                     levels = levels(seu_bb$cluster))

p <- SCpubr::do_AlluvialPlot(sample = seu_hum2,
                             first_group = "cluster",
                             middle_groups = "group",
                             last_group = "singleR_predicted",
                             fill.by = "singleR_predicted", repel = F,
                             colors.use = ph_colors_sr$singleR_predicted)
p

ggsave(filename = "./pdf/singR_pred_human_IVD.pdf", plot = p3, width = 12, height = 8, dpi = 300)
```


## 3.4 integration
```{r, fig.width=8, fig.height=5}
## Young
seu_hYoung <- subset(seu_hum2, group == "Young")
srt_query <- RunKNNMap(srt_query = seu_hYoung, srt_ref = seu_bb, 
                       ref_umap = "umap")
p1 <- ProjectionPlot(
  srt_query = srt_query, srt_ref = seu_bb, 
  ref_param = list(palcolor = list(ph_colors$cluster), pt.alpha = 0.25, pt.size = 3, raster = T),
  query_group = "cluster", ref_group = "cluster",
  query_param = list(palcolor = list(louvain_colors)), pt.size = 1
)
p1 <- p1 + labs(title = "young")
p1
## Adole
seu_hAdole <- subset(seu_hum2, group == "Adole")
srt_query1 <- RunKNNMap(srt_query = seu_hAdole, srt_ref = seu_bb, 
                       ref_umap = "umap")
p2 <- ProjectionPlot(
  srt_query = srt_query1, srt_ref = seu_bb, 
  ref_param = list(palcolor = list(ph_colors$cluster), pt.alpha = 0.25, pt.size = 3, raster = T),
  query_group = "cluster", ref_group = "cluster",
  query_param = list(palcolor = list(louvain_colors)), pt.size = 1
)
p2 <- p2 + labs(title = "adolescent")

(p1 / p2) + plot_layout(guides = "collect")


ggsave(filename = "./pdf/UMAP_pred_human_IVD.pdf", plot = last_plot(), width = 12, 
       height = 15, dpi = 300)
```



## 3.5 KNN prediction
```{r, fig.height=6, fig.width=5}
seu_hYoung <- RunKNNPredict(
  srt_query = seu_hYoung, srt_ref = seu_bb,
  query_group = "cluster", ref_group = "cluster", filter_lowfreq = 20,
  return_full_distance_matrix = TRUE
)

DefaultAssay(seu_hYoung) <- "RNA"
ht <- CellCorHeatmap(
  srt_query = seu_hYoung, srt_ref = seu_bb,
  query_group = "cluster", ref_group = "cluster",
  nlabel = 3, label_by = "row",
  show_row_names = TRUE, show_column_names = TRUE
)
p1 <- ht$plot

ht <- CellCorHeatmap(
  srt_query = seu_hAdole, srt_ref = seu_bb,
  query_group = "cluster", ref_group = "cluster",
  nlabel = 3, label_by = "row",
  show_row_names = TRUE, show_column_names = TRUE
)
p2 <- ht$plot

p1 / p2
ggsave(filename = "./pdf/KNN_similarity_human_IVD.pdf", plot = last_plot(), width = 7, 
       height = 9, dpi = 300)
```
