---
title: "mIVD_RNAseq"
author: "Jian"
date: '2024-06-15'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message=T, warning=T, cache = FALSE, cache.lazy = FALSE, dev = "png",dpi = 300)
Sys.setenv(LANGUAGE = "en") 
options(stringsAsFactors = F, future.fork.enable = TRUE, future.globals.maxSize = 2000 * 1024^1024, 
        repos = c(CRAN="https://mirrors.tuna.tsinghua.edu.cn/CRAN/"),
        BioC_mirror="https://mirrors.tuna.tsinghua.edu.cn/bioconductor/")
```

# 1. Preparations
## 1.1 prepare the envs and path
```{r}
rm(list = ls())
gc()
knitr::opts_knit$set(root.dir = "~/project/mouse_IVD/")
```



# 2. RNA-seq
## 2.1 prepare data
```{r, fig.height=15, fig.width=15}
exprs_gene <- readxl::read_xlsx("~/project/RNAseq/exprs_mIVD.xlsx", sheet = 1)
exprs_gene <- na.omit(exprs_gene)
exprs_gene <- exprs_gene[!duplicated(exprs_gene$`Gene Symbol`), ]
exprs_gene$`Gene Symbol` <- gsub("[']", "", exprs_gene$`Gene Symbol`)
exprs_gene <- as.data.frame(exprs_gene)
rownames(exprs_gene) <- exprs_gene$`Gene Symbol` 

exprs_tpm <- exprs_gene[, c(7:18)]
exprs_fpkm <- exprs_gene[, c(19:30)]
exprs_count <- exprs_gene[, c(31:42)]

meta <- reshape2::colsplit(colnames(exprs_fpkm), " ", c(1,2))
meta <- reshape2::colsplit(meta$`1`, "_", c("sample", "specie", "age"))
meta$group <- substr(meta$sample, 1, 1)
meta$rep <- substr(meta$sample, 2, 3)
rownames(meta) <- meta$sample
colnames(exprs_fpkm) <- colnames(exprs_tpm) <- colnames(exprs_count) <- rownames(meta) 
```  


## 2.2 visulize gene expression
```{r}
genes <- c("Mmp1a","Mmp9","Mmp13","Adamts4","Adamts5", # ECM homeostasis 
           "Tnf", "Il1a",          # Inflammation
           "Vdr", "Pik3ca","Pik3cb", "Akt1", "Akt2", "Akt3", "Mtor",# autophagy
           "Map1lc3a","Map1lc3b","Becn1", # autophagy
           "Bax", "Casp3", "Bcl2", # apoptosis
           "Nox1","Duox1","Duox2","Duoxa1","Duoxa2","Cox5a","Cox5b","Sod1","Sod2",# Oxidative stress 
           "Piezo1","Piezo2", "Trpv1", "Trpv4", "Yap1", "Rock1", "Rock2")
genes%in% rownames(exprs_tpm)
data_gene1 <- exprs_tpm[genes,]

data_gene <- log2(data_gene1+1)
data_gene <- as.data.frame(t(data_gene))
data_gene$group <- meta$group
data_gene <- reshape2::melt(data_gene, id = "group")
data_gene$group <- factor(data_gene$group, levels = c("H","F"))


# choose the test method
test_df <- data.frame(row.names = "gene", pval_norm_H = 0.05, pval_norm_F = 0.05, pval_var = 0.05)
for (i in genes) {
 norm_H <- with(subset(data_gene, variable == i), shapiro.test(value[group == "H"]))
 norm_F <- with(subset(data_gene, variable == i), shapiro.test(value[group == "F"]))
 res.ftest <- var.test(value ~ group, data = subset(data_gene, variable == i))
 test_tmp <- data.frame(row.names = i, pval_norm_H = norm_H$p.value,  pval_norm_F = norm_F$p.value, 
                        pval_var = res.ftest$p.value)
 test_df <- rbind(test_df, test_tmp)
}


theme_set(cowplot::theme_cowplot())
library(ggpubr)
p.list <- list()
for (gene in genes) {
  if (test_df[gene, ]$sig == FALSE) {
    p.list[[gene]] <- ggboxplot(subset(data_gene, variable == gene), x = "group", y = "value", fill = "group",
              palette = c('#a5dff9','#FFBC42'), add = "point", add.params = list(size=2, shape= 21)) + 
      labs( y = paste0("Log2(TPM+1) of ", gene), x = "")+
      theme(legend.position = "none") + 
      scale_y_continuous(expand = expansion(mult = c(0, .1))) +
      stat_compare_means(comparisons = list(c("H","F")), method = "t.test", label = "p.format")
  } else {
    p.list[[gene]] <- ggboxplot(subset(data_gene, variable == gene), x = "group", y = "value", fill = "group",
              palette = c('#a5dff9','#FFBC42'), add = "point", add.params = list(size=2, shape= 5)) + 
      labs( y = paste0("Log2(TPM+1) of ", gene), x = "")+
      theme(legend.position = "none") + 
      scale_y_continuous(expand = expansion(mult = c(0, .1))) +
      stat_compare_means(comparisons = list(c("H","F")), method = "wilcox.test", label = "p.format")
  }
}

library(gridExtra)
library(ggplotify)
p_genelist <- do.call("grid.arrange", c(p.list, ncol = 10))
p_genelist <- as.ggplot(p_genelist)
p_genelist
```


## 2.3 hclust and PCA
```{r, fig.width=10, fig.height=8}
library(ggplot2)
theme_set(cowplot::theme_cowplot())

# 1. compare sequencing depth
data <- data.frame(expression=c(as.matrix(exprs_tpm)),sample=rep(colnames(exprs_tpm),each=nrow(exprs_tpm)))
head(data)
p <- ggplot(data = data,aes(x=sample,y=log2(expression+1),fill=sample))
p1 <- p + geom_boxplot()+ 
  theme(axis.text.x = element_text(angle = 90))+ 
  scale_x_discrete(limits = rev(unique(data$sample)))+
  scale_fill_manual(values = c(rep('#FFBC42',6), rep('#a5dff9',6)))+
  xlab(NULL) + ylab("log2(TPM+1)")
p1



# 2. pca
library(DESeq2)
meta$group <- factor(meta$group, levels = c("H", "F"))

exprs_count_int <- floor(exprs_count)
dds <- DESeqDataSetFromMatrix(countData = exprs_count_int, colData = meta, design = ~ group)
keep <- rowSums(counts(dds) >= 10) >= 3
dds <- dds[keep, ] 
dim(dds)

## 2.1 normalization
library(factoextra)
rld <- vst(dds, blind = T) 
sampleDists <- dist(t(assay(rld)))
res1 <- hcut(sampleDists, k = 2, stand = FALSE,hc_method ="average" ) 
p2 <- fviz_dend(res1,
          rect_fill = T,
          cex = 1,
          color_labels_by_k=T,
          horiz=T) + scale_color_manual(values = c('#FFBC42','#a5dff9'))
p2

exprSet_new <- assay(rld) 
str(exprSet_new)

library(ggplot2)

rld <- vst(dds, blind = F) 
p3 <- plotPCA(rld, intgroup=c('group'))+
  theme_bw() + scale_color_manual(values = c('#a5dff9','#FFBC42'))

(p1 / p3) | p2
ggsave(filename = "./pdf/RNAseq_1.pdf", plot = last_plot(), width = 8, height = 10, dpi = 300)
```

## 2.4 DEseq2
```{r}
library(DESeq2)
colnames(exprs_count)
coldata <- data.frame(group = factor(rep(c('F', 'H'), each = 6), levels = c('H', 'F')))

exprs_count <- floor(exprs_count)
dds <- DESeqDataSetFromMatrix(countData = exprs_count, colData = coldata, design= ~group)
dds1 <- DESeq(dds, fitType = 'mean', minReplicatesForReplace = 7, parallel = FALSE)
res <- results(dds1, contrast = c('group', 'F', 'H'))

res1 <- data.frame(res, stringsAsFactors = FALSE, check.names = FALSE)
res1$gene <- rownames(res1)
res1 <- res1[order(res1$padj, res1$log2FoldChange, decreasing = c(FALSE, TRUE)), ]

res1[which(res1$log2FoldChange >= 1 & res1$padj < 0.01),'sig'] <- 'up'
res1[which(res1$log2FoldChange <= -1 & res1$padj < 0.01),'sig'] <- 'down'
res1[which(abs(res1$log2FoldChange) <= 1 | res1$padj >= 0.01),'sig'] <- 'none'

res1_select <- subset(res1, sig %in% c('up', 'down'))
table(res1_select$sig)
write.csv(res1_select, file = "./RNAseq/DEGs_sig.csv")
res1_select <- read.csv("./RNAseq/DEGs_sig.csv", row.names = 1)
```

## 2.5 enrichment
```{r, fig.width=20, fig.height=6}
library(clusterProfiler)
library(org.Mm.eg.db)
library(enrichplot)
library(aplot)

enrich_down <-  enrichGO(gene          = subset(res1_select, sig == "down")$gene[1:200],
                         OrgDb         = org.Mm.eg.db,
                         keyType       = 'SYMBOL',
                         ont           = "BP",
                         pAdjustMethod = "BH",
                         pvalueCutoff  = 0.01,
                         qvalueCutoff  = 0.1)
View(enrich_down@result)
enrich_down_filtered <- enrich_down
exclude <- grep("muscle|myocyte|blood|cardiac|heart|leukocyte|memory|lymphocyte|T cell", enrich_down_filtered@result$Description)
enrich_down_filtered@result <- enrich_down_filtered@result[-exclude, ]
# View(enrich_down_filtered@result)
enrich_down1 <- pairwise_termsim(enrich_down_filtered)
p1 <- treeplot(enrich_down1, showCategory = 30)
p1

enrich_up <-    enrichGO(gene          = subset(res1_select, sig == "up")$gene[1:200],
                         OrgDb         = org.Mm.eg.db,
                         keyType       = 'SYMBOL',
                         ont           = "BP",
                         pAdjustMethod = "BH",
                         pvalueCutoff  = 0.01,
                         qvalueCutoff  = 0.1)
View(enrich_up@result)
enrich_up_filtered <- enrich_up
exclude <- grep("muscle|myocyte|blood|cardiac|heart|leukocyte|memory|lymphocyte|T|B|immune", enrich_up_filtered@result$Description)
enrich_up_filtered@result <- enrich_up_filtered@result[-exclude, ]
# View(enrich_down_filtered@result)
enrich_up1 <- pairwise_termsim(enrich_up_filtered)
p2 <- treeplot(enrich_up1, showCategory = 30)
p2

library(cowplot)
library(patchwork)
p1 | p2 
```


